spec/classes/all_roles_spec.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+# Copyright (c) 2018 The Regents of the University of Michigan.
+# All Rights Reserved. Licensed according to the terms of the Revised
+# BSD License. See LICENSE.txt for details.
+require 'spec_helper'
+require_relative '../support/contexts/with_mocked_nodes'
+
+def puppet_role_name_from(path)
+  path.strip.gsub('/', '::').gsub(%r{^manifests}, 'nebula').gsub(%r{\.pp}, '')
+end
+
+`find manifests/role -name '*.pp'`.each_line do |file_path|
+  role_name = puppet_role_name_from(file_path)
+  describe role_name do
+    on_supported_os.each do |os, os_facts|
+      context "on #{os}" do
+        let(:facts) { os_facts }
+
+        let(:hiera_config) do
+          if role_name.start_with? 'nebula::role::webhost::htvm'
+            'spec/fixtures/hiera/hathitrust_config.yaml'
+          elsif role_name.start_with? 'nebula::role::hathitrust'
+            'spec/fixtures/hiera/hathitrust_config.yaml'
+          elsif role_name.start_with? 'nebula::role::tools_lib'
+            'spec/fixtures/hiera/tools_lib_config.yaml'
+          elsif role_name.start_with? 'nebula::role::chipmunk'
+            'spec/fixtures/hiera/chipmunk_config.yaml'
+          elsif role_name.start_with? 'nebula::role::deb_server'
+            'spec/fixtures/hiera/deb_server_config.yaml'
+          elsif role_name.start_with? 'nebula::role::deploy_host'
+            'spec/fixtures/hiera/named_instances_config.yaml'
+          else
+            'spec/fixtures/hiera/default_config.yaml'
+          end
+        end
+
+        let!(:puppetdb_query) do
+          MockFunction.new('puppetdb_query') do |f|
+            f.stubbed.returns([])
```

Comment:
A number of roles use `puppetdb_query`, but, luckily, they all seem to be happy with `[]`, which, honestly, they probably always should be.

Reply:
Seems reasonable. I think in general I'm trying to avoid adding new uses of `puppetdb_query`, and think about how to refactor existing uses to avoid it. There are a few places where it's difficult to avoid, though.